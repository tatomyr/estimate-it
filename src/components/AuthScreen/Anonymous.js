import React, { Fragment } from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import {
  Button,
  InputGroup,
  InputGroupAddon,
  Input,
  Alert,
} from 'reactstrap'
import FA from 'react-fontawesome'
import * as api from '../../helpers/api'
import { locationType } from '../../helpers/propTypes'

const AlertAccessingEstimate = ({ from }) => ((from
  && from.pathname
  && from.pathname.startsWith('/estimate/')
  && from.pathname !== '/estimate/new'
)
  ? (
    <Alert color="warning">
      Please enter a valid credentials to get access to this estimate
    </Alert>
  )
  : null)

AlertAccessingEstimate.propTypes = {
  from: locationType.isRequired,
}

const Anonymous = ({
  checkCreds,
  from,
}) => (
  <Fragment>
    <AlertAccessingEstimate from={from} />
    <form
      onSubmit={e => {
        e.preventDefault()
        const [dbName, apiKey] = e.target.credentials.value.split(':')
        api.setCreds({ dbName, apiKey, username: e.target.username.value })
        checkCreds()
      }}
    >
      <label htmlFor="username">
        Public Name
      </label>
      <InputGroup>
        <InputGroupAddon addonType="prepend">
          <span className="input-group-text">
            <FA name="user" />
          </span>
        </InputGroupAddon>
        <Input
          name="username"
          id="username"
          required
          autoFocus // eslint-disable-line jsx-a11y/no-autofocus
        />
      </InputGroup>
      <label htmlFor="credentials">
        Access Key
      </label>
      <InputGroup>
        <InputGroupAddon addonType="prepend">
          <span className="input-group-text">
            <FA name="key" />
          </span>
        </InputGroupAddon>
        <Input
          type="password"
          name="credentials"
          id="credentials"
          required
        />
      </InputGroup>
      <Button color="primary" outline>
        Sign In
      </Button>
    </form>
    Or
    <br />
    <Link to="/estimate/new">
      <Button color="primary" outline>
        Guest Session
      </Button>
    </Link>
  </Fragment>
)

Anonymous.propTypes = {
  checkCreds: PropTypes.func.isRequired,
  from: locationType.isRequired,
}

export default Anonymous
