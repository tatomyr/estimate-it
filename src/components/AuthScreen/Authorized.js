import React, { Fragment } from 'react'
import { Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import { Button } from 'reactstrap'
import * as api from '../../helpers/api'
import { locationType } from '../../helpers/propTypes'

const Authorized = ({
  username,
  resetCreds,
  cleanAllEstimates,
  from,
  redirectToReferrer,
  history,
}) => (
  <Fragment>
    {redirectToReferrer && <Redirect to={from} />}
    <div>
      {/* eslint-disable-next-line react/jsx-one-expression-per-line */}
      Welcome, {username}!
    </div>
    <Button
      color="primary"
      outline
      onClick={history.goBack}
    >
      Close
    </Button>
    <Button
      color="danger"
      outline
      onClick={() => {
        api.removeCreds()
        resetCreds()
        cleanAllEstimates()
      }}
    >
      Log Out
    </Button>
  </Fragment>
)

Authorized.propTypes = {
  username: PropTypes.string.isRequired,
  resetCreds: PropTypes.func.isRequired,
  cleanAllEstimates: PropTypes.func.isRequired,
  from: locationType.isRequired,
  redirectToReferrer: PropTypes.bool,
  history: PropTypes.objectOf(PropTypes.any).isRequired,
}

Authorized.defaultProps = {
  redirectToReferrer: false,
}

export default Authorized
