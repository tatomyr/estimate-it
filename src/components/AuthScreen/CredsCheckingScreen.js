export default () => 'Checking permissions. Please waitâ€¦'
