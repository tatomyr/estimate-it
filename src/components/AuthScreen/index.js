import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import AuthScreen from './AuthScreen'
import './styles.css'
import {
  cleanAllEstimates,
  resetCreds,
} from '../../redux/actions'
import {
  checkCreds,
} from '../../redux/actions/async'

const mapStateToProps = ({ creds }) => ({ creds })

const mapDispatchToProps = ({
  checkCreds,
  resetCreds,
  cleanAllEstimates,
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(AuthScreen))
