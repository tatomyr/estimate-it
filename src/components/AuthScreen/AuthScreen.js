import React from 'react'
import Header from '../Header'
import Authorized from './Authorized'
import Anonymous from './Anonymous'
import CredsCheckingScreen from './CredsCheckingScreen'
import { locationType, credsType } from '../../helpers/propTypes'

const AuthScreen = ({
  creds: {
    haveBeenChecked,
    username,
  },
  location: { state },
  ...rest
}) => (
  <div className="overlay auth-screen">
    <Header />
    <div className="panel">
      {(!haveBeenChecked && <CredsCheckingScreen />)
        || (username && (
          <Authorized
            username={username}
            from={state ? state.from : '/'}
            redirectToReferrer={state && state.redirectToReferrer}
            {...rest}
          />))
        || <Anonymous from={state ? state.from : '/'} {...rest} />}
    </div>
  </div>
)

AuthScreen.propTypes = {
  creds: credsType.isRequired,
  location: locationType.isRequired,
}

export default AuthScreen
