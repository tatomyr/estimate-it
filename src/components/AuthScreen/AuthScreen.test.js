/* globals describe, it, expect */

import React from 'react'
import { HashRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import AuthScreen from './AuthScreen'

const commons = {
  location: { pathname: '/', state: { from: '/' } },
  match: { url: '/estimate/new', params: { estimateId: 'new' } },
  checkCreds: () => null,
  resetCreds: () => null,
  cleanAllEstimates: () => null,
  redirect: () => null,
  history: { goBack: () => null },
}

describe('AuthScreen', () => {
  it('renders correctly for authorized user', () => {
    const tree = renderer
      .create(
        <Router>
          <AuthScreen
            creds={{
              haveBeenChecked: true,
              username: 'Test User',
              dbName: 'dbName',
              apiKey: 'apiKey',
            }}
            {...commons}
          />
        </Router> // eslint-disable-line comma-dangle
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly for anonymous user', () => {
    const tree = renderer
      .create(
        <Router>
          <AuthScreen
            creds={{
              haveBeenChecked: true,
              username: '',
              dbName: '',
              apiKey: '',
            }}
            {...commons}
          />
        </Router> // eslint-disable-line comma-dangle
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when checking credentials', () => {
    const tree = renderer
      .create(
        <Router>
          <AuthScreen
            creds={{
              haveBeenChecked: false,
              username: '',
              dbName: '',
              apiKey: '',
            }}
            {...commons}
          />
        </Router> // eslint-disable-line comma-dangle
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
