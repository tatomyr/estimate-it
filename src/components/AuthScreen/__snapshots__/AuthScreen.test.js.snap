// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthScreen renders correctly for anonymous user 1`] = `
<div
  className="overlay auth-screen"
>
  <nav
    className="navbar navbar-expand navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Estimate It!
    </a>
    <ul
      className="ml-auto navbar-nav"
    >
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="https://github.com/tatomyr/estimate-it"
          onClick={[Function]}
        >
          GitHub
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="panel"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="username"
      >
        Public Name
      </label>
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
            <span
              aria-hidden={true}
              className="fa fa-user"
            />
          </span>
        </div>
        <input
          autoFocus={true}
          className="form-control"
          id="username"
          name="username"
          required={true}
          type="text"
        />
      </div>
      <label
        htmlFor="credentials"
      >
        Access Key
      </label>
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
            <span
              aria-hidden={true}
              className="fa fa-key"
            />
          </span>
        </div>
        <input
          className="form-control"
          id="credentials"
          name="credentials"
          required={true}
          type="password"
        />
      </div>
      <button
        className="btn btn-outline-primary"
        onClick={[Function]}
        type={undefined}
      >
        Sign In
      </button>
    </form>
    Or
    <br />
    <a
      href="#/estimate/new"
      onClick={[Function]}
    >
      <button
        className="btn btn-outline-primary"
        onClick={[Function]}
        type={undefined}
      >
        Guest Session
      </button>
    </a>
  </div>
</div>
`;

exports[`AuthScreen renders correctly for authorized user 1`] = `
<div
  className="overlay auth-screen"
>
  <nav
    className="navbar navbar-expand navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Estimate It!
    </a>
    <ul
      className="ml-auto navbar-nav"
    >
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="https://github.com/tatomyr/estimate-it"
          onClick={[Function]}
        >
          GitHub
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="panel"
  >
    <div>
      Welcome, 
      Test User
      !
    </div>
    <button
      className="btn btn-outline-primary"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
    <button
      className="btn btn-outline-danger"
      onClick={[Function]}
      type="button"
    >
      Log Out
    </button>
  </div>
</div>
`;

exports[`AuthScreen renders correctly when checking credentials 1`] = `
<div
  className="overlay auth-screen"
>
  <nav
    className="navbar navbar-expand navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Estimate It!
    </a>
    <ul
      className="ml-auto navbar-nav"
    >
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="https://github.com/tatomyr/estimate-it"
          onClick={[Function]}
        >
          GitHub
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="panel"
  >
    Checking permissions. Please waitâ€¦
  </div>
</div>
`;
